gradio/route_utils.py

```diff
@@ -583,3 +584,40 @@ def starts_with_protocol(string: str) -> bool:
     """
     pattern = r"^[a-zA-Z][a-zA-Z0-9+\-.]*://"
     return re.match(pattern, string) is not None
+
+
+class CustomCORSMiddleware(BaseHTTPMiddleware):
+    async def dispatch(self, request: fastapi.Request, call_next):
+        # Extract host and origin from the request headers
+        host_header: str = request.headers.get("host", "")
+        origin_header: str = (
+            request.headers.get("origin", "")
+            .replace("http://", "")
+            .replace("https://", "")
+        )
+
+        localhost_prefixes = ("localhost", "127.0.0.1", "0.0.0.0")
+
+        # Check if the host starts with localhost and origin is not localhost
```

Comment:
Doesn't this leave `localhost.evilorigin.bad` wide open?

Reply:
Good point, will tighten this

Reply:
Fixed